packages/attribute-registry-process/src/services/readModelService.ts

```diff
@@ -1,5 +1,223 @@
+      },
+      {
+        $project: {
+          data: 1,
+          computedColumn: { $toLower: ["$data.name"] },
+        },
+      },
+      {
+        $sort: { computedColumn: 1 },
+      },
+    ];
+    const data = await attributes
+      .aggregate([...aggregationPipeline, { $skip: offset }, { $limit: limit }])
+      .toArray();
+    const result = z.array(AttributeTmp).safeParse(data.map((d) => d.data));
+    if (!result.success) {
+      logger.error(
+        `Unable to parse attributes items: result ${JSON.stringify(
+          result
+        )} - data ${JSON.stringify(data)} `
+      );
+      throw ErrorTypes.GenericError;
+    }
+    return {
+      results: result.data,
+      totalCount: await getTotalCount(
+        attributes.aggregate([...aggregationPipeline, { $count: "count" }])
+      ),
+    };
+  },
+
+  async getAttributeById(
+    id: string
+  ): Promise<WithMetadata<AttributeTmp> | undefined> {
+    const data = await attributes.findOne(
+      { "data.id": id },
+      { projection: { data: true, metadata: true } }
+    );
+
+    if (data) {
+      const result = z
+        .object({
+          metadata: z.object({ version: z.number() }),
+          data: AttributeTmp,
+        })
+        .safeParse(data);
+
+      if (!result.success) {
+        logger.error(
+          `Unable to parse attribute item: result ${JSON.stringify(
+            result
+          )} - data ${JSON.stringify(data)} `
+        );
+
+        throw ErrorTypes.GenericError;
+      }
+
+      return {
+        data: result.data.data,
+        metadata: { version: result.data.metadata.version },
+      };
+    }
+
+    return undefined;
+  },
+
+  async getAttributeByName(
+    name: string
+  ): Promise<WithMetadata<AttributeTmp> | undefined> {
+    const data = await attributes.findOne(
+      {
+        "data.name": {
+          $regex: `^${name}$$`,
+          $options: "i",
+        },
+      },
+      { projection: { data: true, metadata: true } }
+    );
+
+    if (data) {
+      const result = z
+        .object({
+          metadata: z.object({ version: z.number() }),
+          data: AttributeTmp,
+        })
+        .safeParse(data);
+
+      if (!result.success) {
+        logger.error(
+          `Unable to parse attribute item: result ${JSON.stringify(
+            result
+          )} - data ${JSON.stringify(data)} `
+        );
+
+        throw ErrorTypes.GenericError;
+      }
+
+      return {
+        data: result.data.data,
+        metadata: { version: result.data.metadata.version },
+      };
+    }
+    return undefined;
+  },
+  async getAttributeByOriginAndCode({
+    origin,
+    code,
+  }: {
+    origin: string;
+    code: string;
+  }): Promise<WithMetadata<AttributeTmp> | undefined> {
+    const codeFilter = {
+      "data.code": code,
+    };
+    const originFilter = {
+      "data.origin": origin,
+    };
+    const data = await attributes.findOne(
+      {
+        codeFilter,
+        originFilter,
```

Comment:
does having nested object for filters work?
btw we can just simplify as
```suggestion
        "data.code": code,
        "data.origin": origin,
```

Reply:
> does having nested object for filters work? btw we can just simplify as

Thanks for the heads up, nested object for filters doesn't work!
I have updated the code similarly as you suggested.